### PR TITLE
add support for payment surcharges

### PR DESCRIPTION
**Pull request by nicoanders**
Added support for payment surcharges, but it also could be used to implement surcharges in general.
The payment surcharges defined per country are still missing in this implementation.

### CODE REVIEW

**Comment by seggewiss** [comment]
@nicoanders Thank you for your work. If you rebase it to our master and resolve any conflicts, I see no trouble in releasing it.

File: Components/Order/Factory/DetailFactory.php [hunk 1]
```diff
@@ -50,8 +50,8 @@ public function create(PositionStruct $positionStruct, $isTaxFree)
             throw new InvalidOrderException('No product number was passed.');
         }
 
-        if ($positionStruct->isDiscount() || $positionStruct->isSurcharge()) {
-            return $this->createDiscount($positionStruct);
+        if ($positionStruct->isDiscount()) {
+            return $this->createDiscount($positionStruct, $isTaxFree);
```
Comment (seggewiss) [inline Components/Order/Factory/DetailFactory.php]: After your rebase you no longer check for $positionStruct->isSurcharge(). I don't think this was intentional ;) Also please verify the correct functionality of this pullrequest. @nicoanders

Comment (nicoanders) [reply Components/Order/Factory/DetailFactory.php]: You´re completly right :) sorry... my fault. I resolve it and squash the commits

**Comment by nicoanders** [comment]
@seggewiss Thanks for caring about the PR.
In all of my tests I could not verify that there is some data missing.
I also added some unit test to ensure some base functionalities.
Could you please describe the steps to fail with more details? Especially what are the surcharge values in the according payment method?

**Comment by seggewiss** [comment]
@nicoanders No need to thank me :) 
First to reproduce this error please pull this branch not any forked version into a fresh environment.
1. Create a new backend order as the payment method choose invoice (surcharge of 2,50€)
    -> add any product, customer and create the order
2. Open the regular order overview and click on edit.

What we see:
-Big yellow warning "Customer was deleted or not found"
-For example, these fields are missing: IP-Address, date time, customer and so on

What we have debugged so far:
The Problem seems to appear on the flush. Before the flush, the object is complete and contains all data, after the flush data gets lost somehow.

**Comment by htkassner** [comment]
Hello. Thank you for your contribution. This does not work correctly. Like Sebastian said it does not create the order correctly. Would be nice of you could solve the described issues. Otherwise we can not merge this pull request.

**Comment by nicoanders** [comment]
Hi @htkassner,
I will provide a solution within this week.

**Comment by htkassner** [comment]
We will close this PR for the moment, due to inactivity, but would be happy if you would contribute this again with another PR. Thank you.

**Pull request closed without merging**